test/Constraints/rdar44770297.swift

```diff
@@ -4,8 +4,8 @@ protocol P {
   associatedtype A
 }
 
-func foo<T: P>(_: () throws -> T) -> T.A? { // expected-note {{in call to function 'foo'}}
+func foo<T: P>(_: () throws -> T) -> T.A? { // expected-note {{where 'T' = 'Never'}}
   fatalError()
 }
 
-let _ = foo() {fatalError()} & nil // expected-error {{generic parameter 'T' could not be inferred}}
+let _ = foo() {fatalError()} & nil // expected-error {{global function 'foo' requires that 'Never' conform to 'P'}}
```

Comment:
This is a clear improvement!

However I can't help but wonder about 'Void'. The diagnostic would say "global function 'foo' requires that 'Void' conform to 'P'", which is a true statement, but 'Void' cannot _ever_ conform to 'P'. Same if it's a function type or a metatype. As a follow-up you might want to add some test cases for non-nominal types and see if the diagnostic could be tailored a bit.